src/main/java/com/google/api/generator/gapic/utils/ResourceReferenceUtils.java

```diff
@@ -0,0 +1,49 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.api.generator.gapic.utils;
+
+import java.util.Arrays;
+import java.util.Optional;
+
+public final class ResourceReferenceUtils {
+
+  private static final String SLASH = "/";
+
+  private ResourceReferenceUtils() {}
+
+  public static Optional<String> parseParentPattern(String pattern) {
```

Comment:
This logic was there before, but it does not look accurate, as it makes many different assumption, which may be false.

for examle, if there is a pattern `/a/{b}/c/e/{f}`
when we search for its parent we must check at least 3 possible cases:
```
/a/{b}/c/e/{f}  - child
/a/{b}/c/e      - parent candidate 1
/a/{b}/c        - parent candidate 2 (if 1 did not match anything)
/a/{b}          - parent candidate 3 (if neither 1 nor 2 matched anything)
```

Instead this logic always returns only one candidate. Not sure how many (if any) practical cases we have which the current implementation would not satisfy. I guess it is ok to keep the implementaiton as is for now, but maybe put a comment/request for refactoring it.

Reply:
It's true, but according to https://google.aip.dev/122, this example is not a recommended pattern.
> Resource name components should usually alternate between collection identifiers (example: publishers, books, users) and resource IDs (example: 123, les-miserables, vhugo1802).

Reply:
Correct, but if it was guaranteed they would phrase it as "must" in the AIP. I think it is ok to keep the implementation as is, since it is beyond the scope of this PR, just maybe add a comment about it.